src/io/bnx.rs

```diff
@@ -0,0 +1,1486 @@
+//! Structs amd trait to read files in BNX format.
+//!
+//! Note: This module only supports version v1.2 of the BNX format for now.
+//!
+//! # Example
+//!
+//! ## Reader
+//!
+//! In this example, we parse a BNX file, iterate over its records. and output some statistics.
+//!
+//! ```rust
+//! use bio::io::bnx;
+//!
+//! let path = "tests/resources/valid_input.bnx";
+//! let mut reader = bnx::Reader::from_path(&path).unwrap();
+//!
+//! let mut num_molecules = 0;
+//! let mut num_labels_channel_1 = 0;
+//! let mut num_labels_channel_2 = 0;
+//!
+//! while let Some(Ok(mut record)) = reader.next() {
+//!     num_molecules += 1;
+//!     num_labels_channel_1 += record.channel_1().len() - 1;
+//!     num_labels_channel_2 += record.channel_2().len() - 1;
+//! }
+//!
+//! println!("Number of molecules: {}", num_molecules);
+//! println!("Number of labels in channel 1: {}", num_labels_channel_1);
+//! println!("Number of labels in channel 2: {}", num_labels_channel_2);
+//! ```
+//!
+//! ## Container
+//!
+//! If feasible, we can also build a container over the content of a BNX file.
+//!
+//! ```rust
+//! use bio::io::bnx;
+//!
+//! let path = "tests/resources/valid_input.bnx";
+//! let mut container = bnx::Container::from_path(&path).unwrap();
+//!
+//! println!("BNX header: {:?}", container.header());
+//! println!("BNX label: {:?}", container.labels());
+//! ```
+
+use anyhow::{bail, Context, Result};
+use derive_new::new;
+use itertools::izip;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+
+use crate::io::om_utils::Error;
+use crate::io::om_utils::NextToErr;
+use crate::io::om_utils::Orientation;
+use crate::io::om_utils::ParseToErr;
+
+lazy_static! {
+    /// Regex command to describe valid position lines in BNX.
+    static ref RE_POS_LINE: Regex =
+        Regex::new(r"\A[12]\t([0-9.\t])*\z").unwrap();
+    /// Regex command to describe valid SNR lines in BNX.
+    static ref RE_SNR_LINE: Regex =
+        Regex::new(r"\AQX[12]1\t([0-9.\t])*\z").unwrap();
+    /// Regex command to describe valid intensity lines in BNX.
+    static ref RE_INTENSITY_LINE: Regex =
+        Regex::new(r"\AQX[12]2\t([0-9.\t])*\z").unwrap();
+}
+
+/// A label as given in a BNX file.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Label {
+    pos: f64,
+    snr: f32,
+    intensity: f32,
+}
+
+/// A BNX record.
+#[derive(Clone, Debug, PartialEq)]
```

Comment:
Use derive Getters and CopyGetters and add things like `#[getset(get="pub")]` or `#[getset(get_copy="pub")]` depending on whether the attribute is a primitive or a complex one.